compiler/rustc_lint/src/nonstandard_style.rs

```diff
@@ -489,21 +490,63 @@ declare_lint! {
 declare_lint_pass!(NonUpperCaseGlobals => [NON_UPPER_CASE_GLOBALS]);
 
 impl NonUpperCaseGlobals {
-    fn check_upper_case(cx: &LateContext<'_>, sort: &str, ident: &Ident) {
+    fn check_upper_case(cx: &LateContext<'_>, sort: &str, did: Option<LocalDefId>, ident: &Ident) {
         let name = ident.name.as_str();
         if name.chars().any(|c| c.is_lowercase()) {
             let uc = NonSnakeCase::to_snake_case(name).to_uppercase();
+
             // We cannot provide meaningful suggestions
             // if the characters are in the category of "Lowercase Letter".
             let sub = if *name != uc {
-                NonUpperCaseGlobalSub::Suggestion { span: ident.span, replace: uc }
+                NonUpperCaseGlobalSub::Suggestion { span: ident.span, replace: uc.clone() }
             } else {
                 NonUpperCaseGlobalSub::Label { span: ident.span }
             };
+
+            struct UsageCollector<'a, 'tcx> {
```

Comment:
Impl `LintDiagnostic` *manually* for `NonUpperCaseGlobal` (it can still remain translatable) and move the usage collection into `decorate_lint`. This way, we only collect usages when absolutely necessary.

Reply:
I did something different I called `opt_span_lint` directly, and since we aren't supposed to (which I kind of disagree) I added a `emit_span_lint_lazy` to do that.

Reply:
Sounds good!